File: recipe-calculator/typescript/src/recipeCalculator.ts
export type Recipe = Record<string, number>;

export function scale(recipe: Recipe, factor: number): Recipe {
  if (factor < 0) {
    throw new RangeError('Scale factor must be non-negative.');
  }

  const scaled: Recipe = {};
  for (const [ingredient, quantity] of Object.entries(recipe)) {
    scaled[ingredient] = quantity * factor;
  }
  return scaled;
}

File: recipe-calculator/csharp/src/RecipeCalculator/RecipeCalculator.cs
namespace RecipeCalculator;

public static class RecipeCalculator
{
    public static IReadOnlyDictionary<string, double> Scale(
        IReadOnlyDictionary<string, double> recipe,
        double factor)
    {
        if (factor < 0)
        {
            throw new ArgumentOutOfRangeException(
                nameof(factor),
                "Scale factor must be non-negative.");
        }

        var scaled = new Dictionary<string, double>(recipe.Count);
        foreach (var (ingredient, quantity) in recipe)
        {
            scaled[ingredient] = quantity * factor;
        }
        return scaled;
    }
}
